### PR TITLE
ESLint config: Print config and add tsdoc plugin

### PR DESCRIPTION
**Pull request by tylerbutler**
Part of #9501.

One question that comes up often when we make changes to our lint config is, "what changed?" This applies even when we don't make any changes other than upgrading deps, because the dependency upgrade might include a new rule.

ESLint provides a way to print the config that would apply to a file (`--print-config`), so this PR uses this capability to print out the applied config as a JSON file. As we make changes to the config, we can print out the config again and get a diff to review as part of a PR -- just like we do with API reports for code changes.

This PR also fixes a bug in the strict and recommended configs -- there was a missing dependency that needed to be added.

### CODE REVIEW

File: common/build/eslint-config-fluid/src/file.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+// This file is used only to run `eslint --print-config`. See the README for this project for more information.
```
Comment (markfields) [inline common/build/eslint-config-fluid/src/file.ts:6]: Did you mean to update the README in this PR?  Instructions for how to use `eslint --print-config` to update the baseline file would be great.

File: common/build/eslint-config-fluid/.eslintrc.js [hunk 2]
```diff
@@ -5,15 +5,7 @@
...
-        // "@rushstack/eslint-config/profile/web-app",
-        "eslint:recommended",
-        "plugin:eslint-comments/recommended",
-        "plugin:import/errors",
-        "plugin:import/warnings",
-        "plugin:import/typescript",
-        // "@rushstack/eslint-config/mixins/tsdoc",
-        // "plugin:unicorn/recommended",
-        "plugin:editorconfig/all",
+        "./minimal"
```
Comment (markfields) [inline common/build/eslint-config-fluid/.eslintrc.js:8]: Why?  What does this .eslintrc.js file do?

**Review by markfields — COMMENTED** [review]
Great idea!  I see there are 2 fake ts files checked in, one under /test/, I'm assuming that's because different rules apply to test v. code files.  And there's also 3 different configs.  So should we have 6 baselines checked in for all the combinations?

File: common/build/eslint-config-fluid/src/file.ts (hunk 1 above)
Comment (tylerbutler) [reply common/build/eslint-config-fluid/src/file.ts:6]: Yeah, good catch. I'll update it.

File: common/build/eslint-config-fluid/.eslintrc.js (hunk 2 above)
Comment (tylerbutler) [reply common/build/eslint-config-fluid/.eslintrc.js:8]: It's the config that applies to the local ts files -- but I am explicitly passing a config when I call eslint (see the package.json scripts) so this file can just be deleted. That should help reduce confusion.

**Comment by tylerbutler** [comment]
> Great idea! I see there are 2 fake ts files checked in, one under /test/, I'm assuming that's because different rules apply to test v. code files. And there's also 3 different configs. So should we have 6 baselines checked in for all the combinations?

Per some feedback from @ChumpChief (https://github.com/microsoft/FluidFramework/pull/9489#discussion_r827371148), I'm consolidating the disparate configs into a single config in my next PR. So I would rather just print the default config, which is the one 99% of our packages use today, and then I'll update it with the results of the config changes in my next PR.

File: common/build/eslint-config-fluid/src/test/file.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+/*!
```
Comment (heliocliu) [inline common/build/eslint-config-fluid/src/test/file.ts:1]: do you need both of these dummy files?

**Comment by tylerbutler** [comment]
@markfields I just pushed an update that includes changes to the readme, a second config for test files, and updates CI to fail the build if a config change is made without printing the config. Can you take a look? Thanks!

File: common/build/eslint-config-fluid/README.md [hunk 4]
```diff
@@ -8,4 +8,37 @@ It exports the following shared ESLint configs:
...
+### Tracking lint config changes over time
+
+One question that comes up often when we make changes to our lint config is, "what changed?" This applies even when we
+don't make any changes other than upgrading deps, because the dependency upgrade might include a new rule.
+
+ESLint provides a way to print the config that would apply to a file (`--print-config`), so we use this capability to
+print out the applied config as a JSON file. As we make changes to the config, we can print out the config again and get
+a diff to review as part of a PR -- just like we do with API reports for code changes.
+
+When updating the lint config, run the `build` script for this package and commit the resulting changes.
```
Comment (heliocliu) [inline common/build/eslint-config-fluid/README.md:28]: sounds like this step should be included in the above section on how to make a change

Comment (tylerbutler) [reply common/build/eslint-config-fluid/README.md:28]: Yup, good point. Will fix.

File: common/build/eslint-config-fluid/src/test/file.ts (hunk 3 above)
Comment (heliocliu) [inline common/build/eslint-config-fluid/src/test/file.ts:1]: do you need two dummy files for this? can both just point to the same dummy?

**Review by heliocliu — APPROVED** [review]

File: common/build/eslint-config-fluid/src/test/file.ts (hunk 3 above)
Comment (tylerbutler) [reply common/build/eslint-config-fluid/src/test/file.ts:1]: ESLint overrides are based on glob paths -- so in order to get ESLint to correctly apply the test-only overrides, the file being linted has to match the test override globs. So it either has to be a .spec.ts file or in the /test/ path. I haven't come up with a better solution :(

File: common/build/eslint-config-fluid/scripts/cleanup-printed-configs.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        // Remove the parser property because it's an absolute path and will vary based on the local environment.
+        delete json.parser;
+
+        // Write out the file
+        fs.writeFileSync(filePath, JSON.stringify(json, undefined, 4));
+
+        // Sort the JSON in-place
+        sortJson.overwrite(filePath, { indentSize: 4 });
+    }
+})();
```
Comment (markfields) [inline common/build/eslint-config-fluid/scripts/cleanup-printed-configs.js:31]: No .catch to log to console or anything?

File: tools/pipelines/build-eslint-config-fluid.yml [hunk 6]
```diff
@@ -64,7 +64,7 @@ extends:
     releaseBuildOverride: ${{ parameters.releaseBuildOverride }}
     buildDirectory: common/build/eslint-config-fluid
     tagName: eslint-config-fluid
-    taskBuild: false
+    taskBuild: build
```
Comment (markfields) [inline tools/pipelines/build-eslint-config-fluid.yml:67]: ?

**Review by markfields — COMMENTED** [review]
I can't click Approve from my phone for some reason. But consider it approved.

**Pull request merged**